npm/pkg/dataplane/ipsets/ipsetmanager_linux.go

```diff
@@ -29,153 +67,409 @@ func (iMgr *IPSetManager) resetIPSets() error {
+	for prefixedName := range iMgr.toDeleteCache {
+		iMgr.flushSetInFile(creator, prefixedName)
+	}
+	for prefixedName := range iMgr.toDeleteCache {
+		iMgr.destroySetInFile(creator, prefixedName)
 	}
 
-	for _, setName := range setNames {
-		setName := setName // to appease golint complaints about function literal
-		errorHandlers := []*ioutil.LineErrorHandler{
-			{
-				Definition: ioutil.NewErrorDefinition(setInUseByKernelPattern),
-				Method:     ioutil.SkipLine,
-				Callback: func() {
-					log.Errorf("was going to delete set %s but it is in use by a kernel component", setName)
-					// TODO mark the set as a failure and reconcile what iptables rule or ipset is referring to it
-				},
-			},
+	// create all sets first so we don't try to add a member set to a list if it hasn't been created yet
+	for prefixedName := range iMgr.toAddOrUpdateCache {
+		set := iMgr.setMap[prefixedName]
+		iMgr.createSetInFile(creator, set)
+	}
+
+	// for dirty sets already in the kernel, update members (add members not in the kernel, and delete undesired members in the kernel)
+	iMgr.updateDirtyKernelSets(saveFile, creator)
+
+	// for the remaining dirty sets, add their members to the kernel
+	for prefixedName := range iMgr.toAddOrUpdateCache {
+		set := iMgr.setMap[prefixedName]
+		sectionID := sectionID(addOrUpdateSectionPrefix, prefixedName)
+		if set.Kind == HashSet {
+			for ip := range set.IPPodKey {
+				iMgr.addMemberInFile(creator, set, sectionID, ip)
+			}
+		} else {
+			for _, member := range set.MemberIPSets {
+				iMgr.addMemberInFile(creator, set, sectionID, member.HashedName)
+			}
 		}
-		sectionID := getSectionID(deletionPrefix, setName)
-		hashedSetName := util.GetHashedName(setName)
-		creator.AddLine(sectionID, errorHandlers, util.IpsetDestroyFlag, hashedSetName) // destroy set
 	}
+	return creator
 }
 
-func (iMgr *IPSetManager) handleAddOrUpdates(creator *ioutil.FileCreator, setNames []string) {
-	// create all sets first
-	// error handling:
-	// - abort the create, flush, and add calls if create doesn't work
-	//   Won't abort adding the set to a list. Will need another retry to handle that
-	//   TODO change this behavior?
-	for _, setName := range setNames {
-		set := iMgr.setMap[setName]
-
-		methodFlag := util.IpsetNetHashFlag
-		if set.Kind == ListSet {
-			methodFlag = util.IpsetSetListFlag
-		} else if set.Type == NamedPorts {
-			methodFlag = util.IpsetIPPortHashFlag
-		}
+// updates the creator (adds/deletes members) for dirty sets already in the kernel
+// updates the toAddOrUpdateCache: after calling this function, the cache will only consist of sets to create
+// error handling principal:
+// - if contract with ipset save (or grep) is breaking, salvage what we can, take a snapshot without grep, and log the failure
+// - have a background process for sending/removing snapshots intermittently
+func (iMgr *IPSetManager) updateDirtyKernelSets(saveFile []byte, creator *ioutil.FileCreator) {
```

Comment:
Does this function walk through all ipset information stored in kernel?
If so, is it necessary?
If there are large number of ipsets are programmed in kernel, this cost will be high?

Do we leverage `-exist` option for `add` and `del` to make it simple?
```shell
# from IPSet man page
add SETNAME ADD-ENTRY [ ADD-OPTIONS ]
Add a given entry to the set. If the -exist option is specified, ipset ignores if the entry already added to the set.
del SETNAME DEL-ENTRY [ DEL-OPTIONS ]
Delete an entry from a set. If the -exist option is specified and the entry is not in the set (maybe already expired), then the command is ignored.
```

Reply:
we grep for all npm sets, so the work is linear in the number of sets in NPM. For non-dirty sets, we do skip reading all add lines in the save file, so this won't be too intensive. We wouldn't know what sets to delete without walking through the ipset save file, so `del --exist` couldn't work, and since we're already determining the members to delete for dirty sets, determining the exact members to add is equal work, and we write a smaller restore file compared to looping through all members and running `add --exist`

Reply:
I may misunderstand the code, but in the code, it also  has string match. So the work is not linear to the number of sets. It is linear to the number of sets and its string in each set.

My fundamental question in the code is why we need to walk through all ipsets (read from kernel and store in `saveFile` variable) in every `applyIPSets`  call and derive some actions (add or deletion) even though we maintain a lot of data structures in  `ipsetmanager_linux.go`. Can we leverage those information and apply what we need?

Reply:
So I guess the decision is between these options:

1. read in save file for dirty sets
2. maintain copy of ipsets object from before they became dirty

We're doing the first here. The second would require extra memory (worst case, doubling the memory for the ipset cache before applying) and would require a redesign of the whole ipset manager. 

Thoughts @vakalapa ?

Reply:
This operation is definitely very intensive wrt string matching, even if we skip over the section of some ipsets we will be doing first of line matching. 

We chose not to save deleted members of IPSets to save memory because we are already very constrained on memory in linux and data structures we are maintaining in Ipset manager is just source of truth. This is modelled against k8s philosophy, we have a source of truth (our internal ipset cache), we read the state on the machine and try to reconcile to that state.

Reply:
Also, this approach is robust in the sense that kernel state will hardly deviate from expected state since NPM takes a snapshot of sets to be updated and then applies changes based on that, current IPSM's draw back is that as errors occur and as NPM runs for longer duration, kernel state heavily deviates from what is expected and this results in a cascading affect of issues, 

Hopefully in the future, when we figure out all the errors and have low-probability-for-error controller, then we can tone this approach down. But for initial purposes i feel, even with this compute intensive update, we are VERY fast compared to previous generation.

Wdygt?

Reply:
If you guys think this is a way, I am ok with this approach.

I just wrote my two cents.
Based on my understanding, the goal of this enhancements in ipset in high-level is
1) Compared to v1 NPM, in v2 we do kind of batch process (one shot for one controller event) unlike in v1 NPM (sequentially called exec.run call for each ipset operation) to improve performance.
2) More fined-grained ipset management with reference counts to avoid known issues (e.g., `used in kernel`)
Do I miss something?

I may be wrong and miss some details,  but to me, it looks like that we took a kind of reverse engineering way, which is hard and complex way while we have enough information to derive all information since data plane is based on all information in control plane.
1) We know when and which operations we need to program (e.g., add, deletion, update) based on controller event) 
2) We know what information are programmed from ipset manger data-structures
3) All the references counts for ipsets
So, I think we can achieve the same functionality with those information in much simple way.

About performance, with a lot of ipsets, I am not sure how much we gains since we do string match O(# of ipsets * # of characters per ipset) in every event which needs applyIPset.

Reply:
As we discussed offline, this current design is more secure than alternatives because even if some malicious actor changes the ipsets, NPM will try to reconcile to the desired state and overwrite manual entries. Granted NPM will only do this for "dirty" sets. A backlog item is added to have a background thread "at leisure" should try to reconcile all ipsets to the desired state, this design will make NPM more secure in the long run. 

Also, perf evaluation is planned on this current design to understand at scale what latency this compute intensive task is going to induce in rule programming.